File: day1/solution.rb
file_name = ARGV[0]

left_list = []
right_list = []

File.foreach(file_name) do |line|
  left, right = line.split.map(&:to_i)
  left_list << left
  right_list << right
end

left_list.sort!
right_list.sort!

total_distance = left_list.each.with_index.reduce(0) do |sum, (left, index)|
  sum + (left - right_list[index]).abs
end

puts "part 1: #{total_distance}"

tally = right_list.tally

similarity_score = left_list.sum { |left| (tally[left] || 0) * left }

puts "part 2: #{similarity_score}"

File: day1/solution.ts
import { readFileSync } from "fs"

const fileName = process.argv[2].split("/").pop()

if (!fileName) {
  throw new Error("No file name")
}

console.log(`READING FILE --- ${fileName} --- `)

const file = readFileSync(fileName, "utf-8").trim().split("\n") as string[]

const [leftList, rightList] = file.reduce(
  (acc, line) => {
    const nums = line.split(" ").map(Number).filter((num) => num != 0)
    if (nums.length > 0) {
      acc[0].push(nums[0])
      acc[1].push(nums[1])
    }

    return acc
  },
  [[], []] as [number[], number[]]
)

const leftSorted = leftList.sort((a, b) => a - b)
const rightSorted = rightList.sort((a, b) => a - b)

let totalDistance = leftSorted.reduce((sum, left, i) => {
  return sum + Math.abs(left - rightSorted[i]);
}, 0);

console.log("part 1", totalDistance)

let similarityScore = 0;
let j = 0;

for (const left of leftSorted) {
  while (j < rightSorted.length && rightSorted[j] < left) {
    j++;
  }

  let count = 0;
  while (j < rightSorted.length && rightSorted[j] === left) {
    count++;
    j++;
  }

  similarityScore += count * left;
}

console.log("part 2", similarityScore)
